rasa/core/nlg/response.py

```diff
@@ -25,28 +26,71 @@ def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
         """
         self.responses = responses
 
+    def _matches_filled_slots(
+        self, filled_slots: Dict[Text, Any], response: Dict[Text, Any],
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION)
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            if filled_slots.get(name) != value:
+                return False
+
+        return True
+
     def _responses_for_utter_action(
-        self, utter_action: Text, output_channel: Text
+        self, utter_action: Text, output_channel: Text, filled_slots: Dict[Text, Any]
     ) -> List[Dict[Text, Any]]:
-        """Return array of responses that fit the channel and action."""
-        channel_responses = []
+        """Returns array of responses that fit the channel, action and condition."""
+        conditional_responses = []
         default_responses = []
 
         for response in self.responses[utter_action]:
-            if response.get("channel") == output_channel:
-                channel_responses.append(response)
-            elif not response.get("channel"):
+            if response.get(RESPONSE_CONDITION) is None:
                 default_responses.append(response)
+            else:
+                matched_response = self._matches_filled_slots(
+                    filled_slots=filled_slots, response=response
+                )
+                if matched_response:
+                    conditional_responses.append(response)
```

Comment:
how adapting this and the prior code to your `filter` stuff below so that it's more uniform?

Reply:
Ok will do, I was also thinking of simplifying the if statements by removing the top first 2 if statements and adapting according to the priority order outlined by Chris.